### ISSUE TITLE
Foldable.iterateRight is not sound

### ISSUE THREAD

**Issue by johnynek** [typelevel/cats#1973]
similar to #1716 I think `Foldable.iterateRight` will break if `Eval.always` are always used.

I think the fix is that we must accept an `Iterable[A]` rather than `Iterator[A]` and call `.iterator` internally. Basically add a `Eval.always(iter.iterator).flatMap { ... }` and I think it should be safe.

Of course, we need to demonstrate the problem with a failing test first.

**Comment by LukaJCB** [comment]
Found a test that doesn't work:

```scala
forAll { (fa: List[Int]) =>
  val eval = Foldable.iterateRight(fa, Eval.later(0)) { (a, eb) =>
    Eval.always(a + eb.value)
  }
  eval.value should === (fa.sum)
  //Repeat here so the result is evaluated again
  eval.value should === (fa.sum)
}
```

**Comment by LukaJCB** [comment]
Fixed by #1980
